packages/content-type-reaction/src/Reaction.ts

```diff
@@ -27,41 +27,45 @@ export type Reaction = {
   schema: "unicode" | "shortcode" | "custom";
 };
 
-export type ReactionParameters = Pick<
-  Reaction,
-  "action" | "reference" | "schema"
-> & {
-  encoding: "UTF-8";
-};
-
 export class ReactionCodec implements ContentCodec<Reaction> {
   get contentType(): ContentTypeId {
     return ContentTypeReaction;
   }
 
-  encode(content: Reaction): EncodedContent<ReactionParameters> {
+  encode(reaction: Reaction): EncodedContent {
+    const { action, reference, schema, content } = reaction;
     return {
       type: ContentTypeReaction,
-      parameters: {
-        encoding: "UTF-8",
-        action: content.action,
-        reference: content.reference,
-        schema: content.schema,
-      },
-      content: new TextEncoder().encode(content.content),
+      parameters: {},
+      content: new TextEncoder().encode(
+        JSON.stringify({ action, reference, schema, content }),
+      ),
     };
   }
 
-  decode(content: EncodedContent<ReactionParameters>): Reaction {
-    const { encoding } = content.parameters;
+  decode(content: EncodedContent): Reaction {
+    let text = new TextDecoder().decode(content.content);
+
+    // First try to decode it in the canonical form.
+    try {
+      const reaction = JSON.parse(text);
+      const { action, reference, schema, content } = reaction;
+      return { action, reference, schema, content };
+    } catch (e) {
+      // ignore, fall through to legacy decoding
+    }
+
+    // If that fails, try to decode it in the legacy form.
+    let parameters = content.parameters || {};
```

Comment:
you could keep the `ReactionParameters` type above (minus the `encoding`) and assert that `parameters` here is that type. then you won't have to do a bunch of assertions in the return object.

Reply:
👍